nixUnstable: 2.4pre20210601_5985b8b -> 2.4pre20210707_02dd6bb

jonringer opened the pull request:
###### Motivation for this change
Includes https://github.com/NixOS/nix/pull/4944, which fixes a long standing GC issue.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
nixUnstable: 2.4pre20210601_5985b8b -> 2.4pre20210707_02dd6bb

jonringer commented:
```
https://github.com/NixOS/nixpkgs/pull/129574

4 packages built:
hydra-unstable nix-review nix-update nixFlakes
```

roberth committed:
nixUnstable: patch boehmgc (nix#4944)

As has been done in https://github.com/NixOS/nix/pull/4944

This introduces the boehmgc_nix and boehmgc_nixUnstable attributes
which are useful for external packages that link with Nix and its
boehmgc.

Mindavi commented:
I've built this and it reports a correct version (so it runs as well):

```
./result/bin/nix --version
nix (Nix) 2.4pre20210707_02dd6bb
```

diff LGTM.

roberth commented:
I've added a commit to patch boehmgc, like https://github.com/NixOS/nix/pull/4944 does. That patch is an essential part of the current gc solution.

jonringer commented:
rebuilt system, and it seems to be working fine

mvnetbiz commented:
This fixes evaluating a Nixos configuration that has been crashing for me with nix-build recently.

Ma27 commented:
Please wait with merging. On this version of Nix I'm currently dealing with some weird messages like `error: ca hash mismatch importing path '/nix/store/hvwszfxaaj4r07yki6plhyfvccs5khpf-childprocess-4.1.0.gem';` for derivations that build perfectly fine with Nix 2.3 (on the client & daemon side).

Investigation will (hopefully) follow tomorrow.

jonringer commented:
> Please wait with merging. On this version of Nix I'm currently dealing with some weird messages like `error: ca hash mismatch importing path '/nix/store/hvwszfxaaj4r07yki6plhyfvccs5khpf-childprocess-4.1.0.gem';` for derivations that build perfectly fine with Nix 2.3 (on the client & daemon side).
> 
> Investigation will (hopefully) follow tomorrow.

I don't use any `ca-references`, so I can't speak to those features

r-rmcgibbo commented:
Result of `nixpkgs-review pr 129574` at 51d480aa run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li>boehmgc_nix</li>
    <li>boehmgc_nixUnstable</li>
    <li>hydra-unstable</li>
    <li>nix-update</li>
    <li>nixUnstable</li>
    <li>nixpkgs-review</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `fetchpatch`.
Near pkgs/tools/package-management/nix/default.nix:1:35:
```
  |
1 | { lib, fetchurl, fetchFromGitHub, fetchpatch, callPackage
  |                                   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/tools/package-management/nix/default.nix:73:17:
```
   |
73 |                 ./aws-sdk-cpp-TransferManager-ContentEncoding.patch
   |                 ^
```
</li>
  </ul>
</details>

Ma27 commented:
> I don't use any ca-references, so I can't speak to those features

Now that you say it, I only have `experimental-features = nix-command flakes` in my `nix.conf`. Apart from that, the hashmismatch appeared because `bundix` gave me a hash that caused problems a few minutes later while running `nix build`. Since one of our selling points is that we can trivially instantiate & realize older Nix expressions (and this doesn't seem to be the case under all circumstances anymore), this has to be cleared up IMHO.

It currently seems as reverting https://github.com/NixOS/nix/pull/4849 on my local Nix checkout fixes the problem, so I'll file a report there later today.

Ma27 commented:
Update: this issue seems to only occur when having multiple stores (e.g. on multiple machines for remote-building). A workaround for the issue is `nix --option builders ''`. I'd personally prefer to still wait until it's clear what's causing this, but if others prefer to merge this now to get the GC fix into master, I'm fine with it as well.

jonringer commented:
I'll have my server up and running later today (hopefully), should be able to verify status of remote builds

jonringer commented:
testing...

Atemu commented:
Could https://github.com/NixOS/nix/pull/4995 be included?

jonringer commented:
I was able to run a few reviews using my server as a remote, and it worked fine....

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14591,6 +14591,18 @@ in
 
   boehmgc = callPackage ../development/libraries/boehm-gc { };
   boehmgc_766 = callPackage ../development/libraries/boehm-gc/7.6.6.nix { };
+  boehmgc_nix = pkgs.boehmgc.override {
+    enableLargeConfig = true;
+  };
+  boehmgc_nixUnstable = pkgs.boehmgc_nix.overrideAttrs (drv: {
+    patches = (drv.patches or []) ++ [
+      # Part of the GC solution in https://github.com/NixOS/nix/pull/4944
+      (pkgs.fetchpatch {
+        url = https://github.com/hercules-ci/nix/raw/5c58d84a76d96f269e3ff1e72c9c9ba5f68576af/boehmgc-coroutine-sp-fallback.diff;
+        sha256 = "sha256-JvnWVTlkltmQUs/0qApv/LPZ690UX1/2hEP+LYRwKbI=";
+      })
+    ];
+  });
```
I am not sure if we want this in the top-level. Why not move it into the nix file?

roberth replied:
I maintain Haskell bindings for Nix and exposing it is useful to me. I guess I could expose it via passthru instead.

roberth committed:
boehmgc_{nix,nixUnstable}: Move into {nix,nixUnstable}.passthru

Do not pollute top-level, but do provide allow the derivations to
be accessed for debugging and reuse if necessary.

jonringer commented:
I rebased this, as it was getting stale

Ma27 commented:
Should we backport this? :)

Mindavi commented:
This seems to have caused https://github.com/NixOS/nix/issues/5044
